Rdkit reb

rebadam opened the pull request:
I added three things:
1) to_rdmol often fails, especially for the iffy compounds made with reactions discovery. I added an optional argument (presanitize) that iteratively sanitizes and update the bonds and atomic charges, untill sanitization succeeds.
2) I added a to_image function that makes a 2D image of a molecule. This uses the presanitize by default. It creates an image (svg (default), png, eps, pdf, jpeg), and optionally prints it to file. It works with our RDKit version, and when using a newer version, it uses newer features. I also works from jupyter notebooks.
3) I added a to_reaction_image function that tries to make a nice looking 2D reaction image.
I added some example molecules and a test script here.
[Example.tar.gz](https://github.com/user-attachments/files/17869800/Example.tar.gz)

rebadam committed:
Rosa Bulo (REB) SCMSUITE-- SO102: Added 2D depiction molecules/reactions + improved stability to_rdmol

- to_rdmol now has an argument presanitize, which iteratively sanitized - improved bonding/atomic charges,
  until sanitization succeeds.
- There are the functions to_image and get_reaction_image.

rebadam committed:
Merge branch 'trunk' into rdkitREB

rebadam committed:
Rosa Bulo (REB) SCMSUITE-- SO102: Cleaned up to_image and get_reaction_image.

- to_image now works with svg in JupyterLab

rebadam committed:
Rosa Bulo (REB) SCMSUITE-- SO102: Cleaned up formatting some more

rebadam committed:
Rosa Bulo (REB) SCMSUITE-- SO102: Cleaned file for mypy

dormrod commented on interfaces/molecule/rdkit.py:
```diff
@@ -108,7 +110,11 @@ def from_rdmol(rdkit_mol: "RDKitMol", confid: int = -1, properties: bool = True)
+    sanitize: bool = True,
+    properties: bool = True,
+    assignChirality: bool = False,
+    presanitize: bool = False,
```
Add `presanitize` to doc-string with description - might be worth also mentioning that sanitize must also be `True` for this to apply

dormrod commented on interfaces/molecule/rdkit.py:
```diff
@@ -213,10 +219,11 @@ def plams_to_rd_bonds(bo):
         try:
-            Chem.SanitizeMol(rdmol)
+            if presanitize:
+                rdmol = _presanitize(plams_mol, rdmol)
```
It would be nice to have add unit-test cases which use this method in `unit_tests/test_molecule_interfaces.TestRDKit` (see e.g. `test_to_rdmol_from_rdmol_roundtrip`).

These could take some input molecules that have issues, and check the output is santised as expected.

Do you have some examples for which this santization is useful?

dormrod commented on interfaces/molecule/rdkit.py:
```diff
@@ -1290,3 +1297,602 @@ def canonicalize_mol(mol, inplace=False, **kwargs):
         return ret
+
+
+@requires_optional_package("rdkit")
```
Updated as also requires optional package `pillow`, as users may not be using `amspython`. 

I have also just sorted out the optional dependencies for PLAMS to make them easier to install, so also added the pillow dependency here: https://github.com/SCM-NV/PLAMS/blob/trunk/pyproject.toml#L59

dormrod commented on interfaces/molecule/rdkit.py:
```diff
@@ -1290,3 +1297,602 @@ def canonicalize_mol(mol, inplace=False, **kwargs):
+
+
+@requires_optional_package("rdkit")
+def to_image(mol, remove_hydrogens=True, filename=None, format="svg", as_string=True):
```
(Added type hints)

dormrod commented on interfaces/molecule/rdkit.py:
```diff
@@ -1290,3 +1297,602 @@ def canonicalize_mol(mol, inplace=False, **kwargs):
+    * ``mol`` -- PLAMS Molecule object
+    * ``remove_hydrogens`` -- Wether or not to remove the H-atoms from the image
+    * ``filename`` -- Optional: Name of image file to be created.
+    * ``format`` -- One of "svg", "png", "eps", "pdf", "jpeg"
```
(renamed to `fmt` to avoid clash with python keyword)

dormrod commented on interfaces/molecule/rdkit.py:
```diff
@@ -1290,3 +1297,602 @@ def canonicalize_mol(mol, inplace=False, **kwargs):
+    # Draw the image
+    if classes[format.lower()] is None:
+        # With AMS version of RDKit MolsToGridImage fails for eps (because of paste)
+        img = Draw.MolToImage(rdmol, size=(200, 100))
```
Might be nice to have the size as an option to the function?

dormrod commented on interfaces/molecule/rdkit.py:
```diff
@@ -1290,3 +1297,602 @@ def canonicalize_mol(mol, inplace=False, **kwargs):
+    return img_text
+
+
+def get_reaction_image_pil(reactants, products, format, width=200, height=100, as_string=True):
```
Make these "private" (`_get_reaction_image_pil`) or add to `__init__`?

dormrod committed:
Minor changes and add unit test SO102

dormrod commented on unit_tests/test_molecule_interfaces.py:
```diff
@@ -230,6 +238,38 @@ def test_to_smiles_and_from_smiles_requires_rdkit_package(self, plams_mols, smil
             with pytest.raises(MissingOptionalPackageError):
                 from_smiles(smiles[0])
 
+    def test_to_image_and_get_reaction_image_can_generate_img_files(self, xyz_folder):
```
I have converted the attached example to a unit test.

It's not too fancy, just makes sure the image files can be generated, but then they can be visually inspected too.

dormrod commented:
Looks like a great feature!

Just a couple of overall comments:

1. I tidied up a few little things I spotted as I reviewed this (type-hints, unit-test etc.)
2. It might also be good to add a notebook example to https://github.com/SCM-NV/PLAMS/tree/trunk/examples, or modify one of the existing ones, to show how to use this new functionality?
3. We have a new changelog which we will use at the end of the year to produce the "what's new" functionality - please could you also add a line in there? https://github.com/SCM-NV/PLAMS/blob/trunk/CHANGELOG.md

rebadam committed:
Rosa Bulo (REB) SCMSUITE-- SO102: presanitize update

- Improved presanitize so that it works for the chlorophyll
- Added unittests for presanitize
- Included presanitize into Mattis MoleculeFormats example

rebadam committed:
Rosa Bulo (REB) SCMSUITE-- SO102: Cleaned up this file:

rebadam committed:
Merge branch 'trunk' into rdkitREB

rebadam committed:
Rosa Bulo (REB) SCMSUITE-- SO107: Added an example of 2D plotting with PLAMS

- Added a jupyter notebook example of 2D plotting of molecules and reactions

rebadam committed:
Rosa Bulo (REB) SCMSUITE-- SO107: Reformatted with black and flake8

rebadam committed:
Rosa Bulo (REB) SCMSUITE-- SO107: Made the new RDKit presanitize functionality work with new RDKit versions

rebadam committed:
Merge branch 'trunk' into rdkitREB

rebadam committed:
Rosa Bulo (REB) SCMSUITE-- SO107: Reformatted with black

rebadam committed:
Rosa Bulo (REB) SCMSUITE-- SO107: ... and flake8 changes...
